File: frontend/src/api/invalidations.ts
import { apiClient } from '@/api/client';

export interface InvalidationEvent {
  id: string;
  building_id: string;
  trigger_type: string;
  trigger_id: string | null;
  trigger_description: string;
  affected_type: string;
  affected_id: string;
  impact_reason: string;
  severity: string;
  required_reaction: string;
  status: string;
  resolved_at: string | null;
  resolved_by_id: string | null;
  resolution_note: string | null;
  detected_at: string | null;
  created_at: string | null;
}

export interface InvalidationPendingResponse {
  items: InvalidationEvent[];
  total: number;
}

export const invalidationsApi = {
  getForBuilding: async (
    buildingId: string,
    params?: { status?: string; severity?: string; limit?: number },
  ): Promise<InvalidationEvent[]> => {
    const response = await apiClient.get<InvalidationEvent[]>(`/buildings/${buildingId}/invalidations`, { params });
    return response.data;
  },

  getPending: async (params?: {
    status?: string;
    severity?: string;
    limit?: number;
  }): Promise<InvalidationPendingResponse> => {
    const response = await apiClient.get<InvalidationPendingResponse>('/invalidations/pending', {
      params,
    });
    return response.data;
  },

  acknowledge: async (eventId: string): Promise<InvalidationEvent> => {
    const response = await apiClient.post<InvalidationEvent>(`/invalidations/${eventId}/acknowledge`);
    return response.data;
  },

  resolve: async (eventId: string, resolutionNote: string): Promise<InvalidationEvent> => {
    const response = await apiClient.post<InvalidationEvent>(`/invalidations/${eventId}/resolve`, {
      resolution_note: resolutionNote,
    });
    return response.data;
  },

  executeReaction: async (eventId: string): Promise<Record<string, unknown>> => {
    const response = await apiClient.post<Record<string, unknown>>(`/invalidations/${eventId}/execute-reaction`);
    return response.data;
  },
};

File: backend/app/api/invalidations.py
"""API routes for the invalidation engine."""

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies import require_permission
from app.models.user import User
from app.schemas.invalidation import (
    InvalidationEventRead,
    InvalidationPendingResponse,
    InvalidationResolveRequest,
)
from app.services.invalidation_engine import InvalidationEngine

router = APIRouter()


async def _get_building_or_404(db: AsyncSession, building_id: UUID):
    from app.services.building_service import get_building

    building = await get_building(db, building_id)
    if not building:
        raise HTTPException(status_code=404, detail="Building not found")
    return building


@router.get(
    "/buildings/{building_id}/invalidations",
    response_model=list[InvalidationEventRead],
    tags=["Invalidations"],
)
async def get_building_invalidations(
    building_id: UUID,
    status: str | None = Query("detected", description="Filter by status"),
    severity: str | None = Query(None, description="Filter by severity"),
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
    current_user: User = Depends(require_permission("buildings", "read")),
    db: AsyncSession = Depends(get_db),
):
    """Get invalidation events for a building."""
    await _get_building_or_404(db, building_id)
    engine = InvalidationEngine()
    items, _total = await engine.get_pending_invalidations(
        db,
        building_id=building_id,
        status=status,
        severity=severity,
        limit=limit,
        offset=offset,
    )
    return items


@router.get(
    "/invalidations/pending",
    response_model=InvalidationPendingResponse,
    tags=["Invalidations"],
)
async def get_pending_invalidations(
    status: str | None = Query("detected", description="Filter by status"),
    severity: str | None = Query(None, description="Filter by severity"),
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
    current_user: User = Depends(require_permission("buildings", "read")),
    db: AsyncSession = Depends(get_db),
):
    """Get all pending invalidation events across the organization."""
    engine = InvalidationEngine()
    items, total = await engine.get_pending_invalidations(
        db,
        org_id=current_user.organization_id,
        status=status,
        severity=severity,
        limit=limit,
        offset=offset,
    )
    return {"items": items, "total": total}


@router.post(
    "/invalidations/{event_id}/acknowledge",
    response_model=InvalidationEventRead,
    tags=["Invalidations"],
)
async def acknowledge_invalidation(
    event_id: UUID,
    current_user: User = Depends(require_permission("buildings", "update")),
    db: AsyncSession = Depends(get_db),
):
    """Acknowledge an invalidation event."""
    engine = InvalidationEngine()
    event = await engine.acknowledge_invalidation(db, event_id, current_user.id)
    if event is None:
        raise HTTPException(status_code=404, detail="Invalidation event not found")
    await db.commit()
    return event


@router.post(
    "/invalidations/{event_id}/resolve",
    response_model=InvalidationEventRead,
    tags=["Invalidations"],
)
async def resolve_invalidation(
    event_id: UUID,
    data: InvalidationResolveRequest,
    current_user: User = Depends(require_permission("buildings", "update")),
    db: AsyncSession = Depends(get_db),
):
    """Resolve an invalidation event."""
    engine = InvalidationEngine()
    event = await engine.resolve_invalidation(db, event_id, current_user.id, data.resolution_note)
    if event is None:
        raise HTTPException(status_code=404, detail="Invalidation event not found")
    await db.commit()
    return event


@router.post(
    "/invalidations/{event_id}/execute-reaction",
    tags=["Invalidations"],
)
async def execute_reaction(
    event_id: UUID,
    current_user: User = Depends(require_permission("buildings", "update")),
    db: AsyncSession = Depends(get_db),
):
    """Execute the required reaction for an invalidation event."""
    engine = InvalidationEngine()
    result = await engine.execute_reaction(db, event_id)
    if "error" in result and result.get("error") == "Event not found":
        raise HTTPException(status_code=404, detail="Invalidation event not found")
    await db.commit()
    return result
